hybrasyl/CreatureStatus.cs

```diff
@@ -136,39 +151,38 @@ public class CreatureStatus : ICreatureStatus
 
         public double ElapsedSinceTick => (DateTime.Now - LastTick).TotalSeconds;
 
-        private void _init(Status xmlstatus, Creature target, Castable castable)
+        public CreatureStatus(Status xmlstatus, Creature target, Creature source=null, Castable castable = null,
+            int durationOverride = -1, int tickOverride = -1)
         {
-            XMLStatus = xmlstatus;
             Target = target;
+            Source = source;
+            XMLStatus = xmlstatus;
             Castable = castable;
-
-        }
-        public CreatureStatus(Status xmlstatus, Creature target, Castable castable = null,
-            int durationOverride = -1, int tickOverride = -1)
-        {
-            _init(xmlstatus, target, castable);
             Start = DateTime.Now;
             _durationOverride = durationOverride;
             _tickOverride = tickOverride;
         }
 
-        public CreatureStatus(Status xmlstatus, Creature target, Castable castable = null)
+        public CreatureStatus(Status xmlstatus, Creature target, Creature source=null, Castable castable = null)
```

Comment:
It should not matter who applied the status.

Reply:
Formulas can be used in status damage. There are several examples of formulas that use `$CASTER` values. This ensures they work.

Reply:
I'd like an example of a status that uses a caster based formula. From my understanding is that status formulas should not have any reliance on the specific caster.

Reply:
Statuses are not really implemented 100% yet on the data side, so there is not one currently. There are many examples of valid formulas that would need this information. Are you saying you'd rather have the potential for multiple `NullReferenceExceptions` in damage calculation rather than including an object reference in a status? That seems very strange to me. 

There is no reliance - currently - but we decided to allow formulas in statuses, so, this makes that work correctly with intended results and no surprises. If we only want statuses to have a specific set of formulas or to have no formulas, that is a different question altogether, but it seems like bad behavior to say "You can only use a subset of the standard formula variables here" to XML creators.

Reply:
Reliance on caster variables in status formulas will introduce a slew of complexity that should be accounted for here. If this is the path that you'd like to take, then so be it, but I'm almost guaranteeing that it becomes very messy.

For example, how to you maintain the caster data at the time of spell cast versus just a solid reference to the caster? How does this work when logging off and logging back on? What happens to the status during a leveling cycle and the caster levels up? Or changes equipment? Or changes stats? Or while master buys stats?

This is just a sampling of the variables that need to be accounted for if this is the path you wish to take with this.

Reply:
Fair point about logging on and off. What about evaluating the formula once on application and storing the result for future use? Otherwise we're basically saying that statuses can only have simple damage, which seems pretty limiting.